### ISSUE TITLE
CRI: Implement container log rotation

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#58823]
TL;DR: We want to allow kubelet to rotate the container stdout/stderr logs, and add a CRI call to ask the runtime to reopen the log file.

kubelet decides the log directory structure and passes to the runtime the path where container log should be stored. Given the level of knowledge and control kubelet has, it is currently the best candidate to rotate the logs (vs. runtime or a separate daemon). The detailed evaluation is in the [proposal doc](https://docs.google.com/document/d/1oQe8dFiLln7cGyrRdholMsgogliOtpAzq6-K3068Ncg/edit?usp=sharing).

IMO, the initial version should target a simple policy of size limit per file. We can iterate and add more advanced features based on disk management in the future. The stretch goal for this release will be supporting reading logs (`kubectl logs`) from rotated log files.

As for the docker integration (i.e., `--container-runtime=docker`) that does not fully implement CRI logging, kubelet should skip log rotation.

/cc @kubernetes/sig-node-feature-requests @dchen1107 @Random-Liu @mrunalp @resouer @feiskyer

**Comment by yujuhong** [comment]
The CRI change has been merged. What's left to do is implementing in kubelet, and adding a test for it once we have a CRI runtime supporting the feature.

**Comment by jberkus** [comment]
Hey, if the remaining work for this is going into 1.10, can you please add a priority label?  I also don't see the remaining work on the Features board?  If this isn't rolled into another feature, then you've missed feature freeze for this one ...

**Comment by idvoretskyi** [comment]
@yujuhong can you link this issue to the one in a features repo?

**Comment by yujuhong** [comment]
> @yujuhong can you link this issue to the one in a features repo?

@jberkus @idvoretskyi a lot of these are ongoing work in the past few releases to bridge the gap between docker and non-docker runtimes using CRI. They are not used with docker (the only runtime with in-tree support), and are in general not user facing. If all milestone issues must link to a feature, hope this would help: https://github.com/kubernetes/features/issues/552

**Comment by yujuhong** [comment]
@Random-Liu is going to own the issue unless we find someone else with more bandwidth this release.
